saleor/order/actions.py

```diff
@@ -1361,37 +1380,51 @@ def _process_refund(
     refund_amount = _calculate_refund_amount(
         order_lines_to_refund, fulfillment_lines_to_refund, lines_to_refund
     )
+
+    # Happens when neither payments_to_refund nor amount_to_refund were specified.
     if amount is None:
         amount = refund_amount
         # we take into consideration the shipping costs only when amount is not
         # provided.
         if refund_shipping_costs:
             amount += order.shipping_price_gross_amount
+        # At this point there can be only one payment.
+        payments[0]["amount"] = min(payments[0]["payment"].captured_amount, amount)
```

Comment:
Not sure of this. (probably I am missing something) Why do we do this? What in the case,  when we are not able to include shipping_price to amount of first payment because captured_amount of first payment is less than amount+shipping_price?
`refund_shipping_costs` was the flag that determines if we want to refund shipping costs for payment, when we have a single payment. Shouldn't we operate on `item["include_shipping_costs"]` and drop this arguments?

Reply:
Still, I don't understand it. `proccess_refund` should call refund action for each payment. The total  amount of refund should be taken from automatically calculated refund OR from `payment['amount']` + `payment['include_shipping_costs']` determined for each payment.
In case when we don't have `payment['amount']` value, then we should use `min(payment.captured,total_amound_of_refund)`. The cases which we need to cover:

- _user doesn't provide amount of returned value but needs to provide a list of paymentIDs in proper order, which will be used to call refund. (In case when amount to return is greater than amount from single payment. For amount lower than amount of single payment the list can have a single paymentID).  Saleor calculates refund amount, call refund for max amount of payment and move to next payment from provided list_ -  This will use automatically calculated total of refund and it will iterate over a payment and process a refund one by one until we will cover `total of refund`. 

- _user provides amount to refund for each payment. 
example when order has two payments to refund:
15 USD for 1st payment
10 USD for 2nd payment_ - we call try_refund for 'amount' defined for each `payment' object
- _user provides amount only for part of refunded payments - [{paymentID: "ID1"}, {"paymentID: "ID2", amount: 15}]
First call refund for payments with defined amount, then call refund for payments which don't have provided amount_ - `_process_refund` should call refund for objects with determined `amount` value, then we should call refund for the rest of the payments, where amount to refund is == `total_amount_of_refund - sum(payments['amount'] and is not None)`

Reply:
All the cases described here don't apply to the line above. When a list of payments is provided we include the shipping costs in this line:
https://github.com/mirumee/saleor/pull/8165/files#diff-45007f7469800f0df7ae204bece8aa9303d0b7261e52ac1fdd9a5d710d089368R916

I'd consider this as a transitioning step that we can refactor later.